driver/src/main/com/mongodb/client/gridfs/GridFSBucketImpl.java

```diff
@@ -0,0 +1,321 @@
+        this.chunksCollection = notNull("chunksCollection", chunksCollection);
+    }
+
+    @Override
+    public int getChunkSizeBytes() {
+        return chunkSizeBytes;
+    }
+
+    @Override
+    public ReadPreference getReadPreference() {
+        return readPreference;
+    }
+
+    @Override
+    public WriteConcern getWriteConcern() {
+        return writeConcern;
+    }
+
+    @Override
+    public GridFSBucket withChunkSizeBytes(final int chunkSizeBytes) {
+        return new GridFSBucketImpl(database, bucketName, chunkSizeBytes, codecRegistry, readPreference, writeConcern, filesCollection,
+                chunksCollection, checkedIndexes);
+    }
+
+    @Override
+    public GridFSBucket withReadPreference(final ReadPreference readPreference) {
+        return new GridFSBucketImpl(database, bucketName, chunkSizeBytes, codecRegistry, readPreference, writeConcern, filesCollection,
+                chunksCollection, checkedIndexes);
+    }
+
+    @Override
+    public GridFSBucket withWriteConcern(final WriteConcern writeConcern) {
+        return new GridFSBucketImpl(database, bucketName, chunkSizeBytes, codecRegistry, readPreference, writeConcern, filesCollection,
+                chunksCollection, checkedIndexes);
+    }
+
+    @Override
+    public GridFSUploadStream openUploadStream(final String filename) {
+        return openUploadStream(filename, new GridFSUploadOptions());
+    }
+
+    @Override
+    public GridFSUploadStream openUploadStream(final String filename, final GridFSUploadOptions options) {
+        int chunkSize = options.getChunkSizeBytes() == null ? chunkSizeBytes : options.getChunkSizeBytes();
+        Bson metadata = options.getMetadata() == null ? new BsonDocument() : options.getMetadata();
+        BsonDocument metadataBsonDocument = metadata.toBsonDocument(BsonDocument.class, codecRegistry);
+        checkCreateIndex();
+        return new GridFSUploadStreamImpl(filesCollection, chunksCollection, new ObjectId(), filename, chunkSize, metadataBsonDocument);
+    }
+
+    @Override
+    public ObjectId uploadFromStream(final String filename, final InputStream source) {
+        return uploadFromStream(filename, source, new GridFSUploadOptions());
+    }
+
+    @Override
+    public ObjectId uploadFromStream(final String filename, final InputStream source, final GridFSUploadOptions options) {
+        GridFSUploadStream uploadStream = openUploadStream(filename, options);
+        int chunkSize = options.getChunkSizeBytes() == null ? chunkSizeBytes : options.getChunkSizeBytes();
+        byte[] buffer = new byte[chunkSize];
+        int len;
+        try {
+            while ((len = source.read(buffer)) != -1) {
+                uploadStream.write(buffer, 0, len);
+            }
+        } catch (IOException e) {
+            throw new MongoGridFSException("IO Exception when reading from the InputStream", e);
+        } finally {
+            uploadStream.close();
```

Comment:
No, if the finally block throws the original exception is replaced before propagation.  The original should be thrown so there needs to be a try/catch in the finally block itself, unfortunately.